### PR TITLE
RM; pass solid pressure to density

### PR DESCRIPTION
**Pull request by endJunction**
Followup of #2996 

Required for solid pressure dependent density.

1. [x] Feature description was added to the [changelog](https://github.com/ufz/ogs/wiki/Release-notes-6.3.1)
2. [x] Tests covering your feature were added? ctest with swelling was updated to include the new dependency
3. [x] Any new feature or behavior change was documented?

### CODE REVIEW

**Comment by endJunction** [comment]
exponential_solid_density

dashed lines - constant solid density
solid lines - exponential solid density.

All but the dry densities plots are identical.
See benchmark itself for details. Short: mechanical load phase in the beginning, desaturation at the end with swelling happening.

File: ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h [hunk 1]
```diff
@@ -803,6 +800,15 @@ void RichardsMechanicsLocalAssembler<ShapeFunctionDisplacement,
                 displacement_index, displacement_index)
             .noalias() += B.transpose() * C * B * w;
 
+        auto const phi = _ip_data[ip].porosity;
```
Comment (TomFischer) [inline ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h:803]: Why not:
```suggestion
        auto const& phi = _ip_data[ip].porosity;
```
?

File: ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h [hunk 2]
```diff
@@ -1535,10 +1541,15 @@ void RichardsMechanicsLocalAssembler<ShapeFunctionDisplacement,
 
         GlobalDimMatrixType const K_over_mu = k_rel * K_intrinsic / mu;
 
+        auto const phi = _ip_data[ip].porosity;
```
Comment (TomFischer) [inline ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h:1544]: Same as above, why not
```suggestion
        auto const& phi = _ip_data[ip].porosity;
```

File: ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h (hunk 1 above)
Comment (endJunction) [reply ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h:803]: Is a double.

File: ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h [hunk 3]
```diff
@@ -803,6 +800,15 @@ void RichardsMechanicsLocalAssembler<ShapeFunctionDisplacement,
                 displacement_index, displacement_index)
             .noalias() += B.transpose() * C * B * w;
 
+        auto const phi = _ip_data[ip].porosity;
+        double const p_FR = -chi_S_L * p_cap_ip;
+        // p_SR
+        variables[static_cast<int>(MPL::Variable::solid_grain_pressure)] =
+            p_FR - (sigma_eff + sigma_sw).dot(identity2) / (3 * (1 - phi));
```
Comment (wenqing) [inline ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h:807]:
```suggestion
            p_FR - std::max(0.0, (sigma_eff + sigma_sw).dot(identity2) / (3 * (1 - phi)));
```

Comment (wenqing) [reply ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h:807]: or 
```
std::max(0.0, p_FR - (sigma_eff + sigma_sw).dot(identity2) / (3 * (1 - phi)));
```

Comment (endJunction) [reply ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h:807]: why?

Comment (wenqing) [reply ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h:807]: I do know what it is exactly the density model. If the density model describes the density change only under pressure, or compression, and no change in extension, all three principle stresses must be negative. If it is not the case, your implementation is roughly good, which means we do not care compressed or extended in directions but with a dominated status.

Comment (endJunction) [reply ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h:807]: The density model in the ctest is exponential one describing the change in both directions, compression and extension.

File: Tests/Data/RichardsMechanics/deformation_dependent_porosity_swelling.prj [hunk 4]
```diff
@@ -72,8 +72,16 @@
...
+                            <type>Exponential</type>
+                            <reference_value>1</reference_value>
+                            <exponent>
+                                <variable_name>solid_grain_pressure</variable_name>
+                                <reference_condition>0</reference_condition>
+                                <!-- beta_SR = (1 - alpha)/K_S. K_S = E/(3(1-2 nu)) -->
+                                <!-- K_S = 1/(3*(1-2*0.3)) = 5/6. -->
+                                <!-- beta_SR = (1 - 11/12)/K_S = 0.1 -->
+                                <factor>0.1</factor>
+                            </exponent>
```
Comment (endJunction) [inline Tests/Data/RichardsMechanics/deformation_dependent_porosity_swelling.prj:84]: @wenqing the used density model.

Comment (wenqing) [reply Tests/Data/RichardsMechanics/deformation_dependent_porosity_swelling.prj:84]: I see.

**Pull request merged**